lib/cudnn/convolution.jl

```diff
@@ -0,0 +1,259 @@
+                cudnnFindConvolutionForwardAlgorithmEx(handle(),xDesc,x,wDesc,w,convDesc,yDesc,y,requestedAlgoCount,returnedAlgoCount,perfResults,workspace,sizeof(workspace))
+            end
+            cudnnConvolutionAlgoPerfChoose(perfResults, returnedAlgoCount[1])
+        end
+    end
+end
+
+const cudnnConvolutionBwdDataAlgoPerfCache = Dict{Tuple,cudnnConvolutionBwdDataAlgoPerf_t}()
+function cudnnConvolutionBwdDataAlgoPerf(wDesc, w, dyDesc, dy, convDesc, dxDesc, dx)
+    get!(cudnnConvolutionBwdDataAlgoPerfCache, (wDesc, dyDesc, convDesc)) do 
+        requestedAlgoCount = Int(CUDNN_CONVOLUTION_BWD_DATA_ALGO_COUNT)
+        returnedAlgoCount = Cint[0]
+        perfResults = Array{cudnnConvolutionBwdDataAlgoPerf_t}(undef,requestedAlgoCount)
+        @workspace size=cudnnFindConvolutionAlgorithmWorkspaceSize(dx) workspace->begin
+            cudnnFindConvolutionBackwardDataAlgorithmEx(handle(),wDesc,w,dyDesc,dy,convDesc,dxDesc,dx,requestedAlgoCount,returnedAlgoCount,perfResults,workspace,sizeof(workspace))
+        end
+        cudnnConvolutionAlgoPerfChoose(perfResults, returnedAlgoCount[1])
+    end
+end
+
+const cudnnConvolutionBwdFilterAlgoPerfCache = Dict{Tuple,cudnnConvolutionBwdFilterAlgoPerf_t}()
+function cudnnConvolutionBwdFilterAlgoPerf(xDesc, x, dyDesc, dy, convDesc, dwDesc, dw)
+    get!(cudnnConvolutionBwdFilterAlgoPerfCache, (xDesc, dyDesc, convDesc)) do 
+        requestedAlgoCount = Int(CUDNN_CONVOLUTION_BWD_FILTER_ALGO_COUNT)
+        returnedAlgoCount = Cint[0]
+        perfResults = Array{cudnnConvolutionBwdFilterAlgoPerf_t}(undef,requestedAlgoCount)
+        @workspace size=cudnnFindConvolutionAlgorithmWorkspaceSize(x) workspace->begin
+            cudnnFindConvolutionBackwardFilterAlgorithmEx(handle(),xDesc,x,dyDesc,dy,convDesc,dwDesc,dw,requestedAlgoCount,returnedAlgoCount,perfResults,workspace,sizeof(workspace))
+        end
+        cudnnConvolutionAlgoPerfChoose(perfResults, returnedAlgoCount[1])
+    end
+end
+
+
+# Return algorithm with best memory that is within 10% of best time
+function cudnnConvolutionAlgoPerfChoose(ps, n)
+    (ibest,mbest,tbest) = (0,Inf,Inf)
+    for i in 1:n
+        # These metrics are written in a sorted fashion where the first element has the lowest compute time.
+        if ps[i].status == CUDNN_STATUS_SUCCESS && ps[i].memory < mbest && ps[i].time < tbest * 1.1
+            (ibest,mbest,tbest) = (i,ps[i].memory,ps[i].time)
+        end
+    end
+    if ibest == 0
+        @warn "No valid algorithm found, probably bad params for convolution." maxlog=1
+        ibest = findfirst(p->p.algo==0, ps)
+        ibest === nothing && error("Cannot find backup algorithm for convolution, giving up.")
+    end
+    return ps[ibest]
+end
+
+
+# Allocate the maximum reasonable amount of memory for algorithm discovery
+function cudnnFindConvolutionAlgorithmWorkspaceSize(x)
+    gpufree = Mem.info()[1] + CUDA.cached_memory()
+    min(gpufree ÷ 10, sizeof(x) * 100)
```

Comment:
100 seems a bit aggressive, it might mean we start going to slower algos as we get more memory constrained, when tensors should have been freed. Maybe @maleadt can clarify if my thinking is incorrect here

Reply:
`Mem.available_memory()` instead of `Mem.info()[1]`.

100x seems aggressive indeed; is there any guide in the documentation, or how did you come up with that?